Add index to tool_crawler_url url field #113

kristian-94 opened the pull request:
This is for #113

brendanheywood commented on db/install.xml:
```diff
@@ -38,6 +38,7 @@
         <INDEX NAME="courseid" UNIQUE="false" FIELDS="courseid"/>
         <INDEX NAME="httpcode" UNIQUE="false" FIELDS="httpcode"/>
         <INDEX NAME="priority_needscrawl" UNIQUE="false" FIELDS="needscrawl, priority"/>
+        <INDEX NAME="url" UNIQUE="false" FIELDS="url"/>
```
should this be unique?

kristian-94 replied:
I'll resolve this convo, not applicable anymore since we can't just add an index here.

brendanheywood commented:
Just note for future, this needs a much bigger change to do problem to deal with index length curve balls

brendanheywood commented on db/upgrade.php:
```diff
@@ -109,5 +109,51 @@ function xmldb_tool_crawler_upgrade($oldversion) {
+        if ($dbman->field_exists($table, $field)) {
+            $dbman->rename_field($table, $field, 'externalurl');
+        }
+
```
I think you will need some data migration too, to back fill all the new columns with legit data

kristian-94 commented:
An update:
I have implemented the persistent API, and made sure that each update and insert to the tool_crawler_url table utilizes it (no direct inserts).

I've changed the upgrade script to include the new urlhash field and populate it. Queries that selected on url have been changed to now select on urlhash.

I have a unit test that tests that the url hash field is correct when creating and updating records.

I think this is getting pretty close now. It will need some end to end testing as the next step.

brendanheywood commented on classes/robot/crawler.php:
```diff
@@ -347,7 +347,7 @@ public function mark_for_crawl($baseurl, $url, $courseid = null, $priority = TOO
 
         // Find the current node in the queue.
-        $node = url::get_record(['url' => $url]);
+        $node = url::get_record(['urlhash' => md5($url)]);
```
can you please refactor a shared hash util so that we can change the url -> hash function in a single place

kristian-94 replied:
Done in the latest commit

brendanheywood commented on classes/local/url.php:
```diff
@@ -0,0 +1,239 @@
+     * @return string the hashed url
+     */
+    public static function hash_url($url) {
+        return md5($url);
```
given how broken md5 is probably worth swapping to sha1, even though this isn't a crypto issue it can't hurt

kristian-94 replied:
Switched

brendanheywood commented on classes/task/robot_cleanup.php:
```diff
@@ -78,6 +78,7 @@ public function execute($currenttime = null) {
             $where = 'lastcrawled <= :currenttime
                   AND lastcrawled <= :lastcrawlfinished
                   AND lastcrawled <= :expiredate';
+            // might want to move this to persistent
```
still valid comments?

brendanheywood commented on db/upgrade.php:
```diff
@@ -109,5 +111,69 @@ function xmldb_tool_crawler_upgrade($oldversion) {
+                }
+                // Before indexing, populate the urlhash field with hashes of the currently existing urls.
+                core_php_time_limit::raise();
+                $urlrecords = $DB->get_recordset('tool_crawler_url');
```
this will be crazy slow to churn through all the urls. Can this be done in sql?

SELECT encode(digest('<root/>'::bytea, 'sha1'), u.url)

brendanheywood commented on version.php:
```diff
@@ -27,7 +27,7 @@
 
-$plugin->version   = 2020030500;        // The current plugin version (Date: YYYYMMDDXX)
+$plugin->version   = 2020030501;        // The current plugin version (Date: YYYYMMDDXX)
 $plugin->release   = 2020030500;        // The current plugin version (Date: YYYYMMDDXX)
```
bump release too please
